### DISCUSSION TITLE
multi-sourced download with checksum

### DISCUSSION THREAD

**Discussion by nyurik** [chridou/condow#35] [Q&A]
Hi, I need to download a file from multiple mirrors (sources), possibly in multiple streams, with retries (similar to  [aria2c](https://aria2.github.io/manual/en/html/aria2c.html), but as a lib). Once downloaded, I need to calculate a checksum (usually md5).  Your lib seems to target exactly this, but it seems it is incomplete for http sources per #26, and the checksum is also missing (?).  Could you suggest a best path to get started, and perhaps I could collaborate on this lib too?

WRT checksum -- I could compute MD5 after the file is fully downloaded, but that could significantly slow down overall performance of my app, thus ideally it should be possible to compute it as soon as the next sequential blob becomes available. In the best case if the last blob becomes available last, it will delay only by the time of iterating over the last blob because prior data would have been done by then. Ability to stream while downloading might also benefit various movie streaming apps (view while downloading).

P.S. this is more of a question more suitable for the discussion - could you enable them?

**Comment by chridou** [comment]
Hi! This is an interesting topic and a missing feature. I have enabled discussions.
My first thought would be to use the [PartStream](https://github.com/chridou/condow/blob/main/condow_core/src/streams/part_stream.rs#L53) which is ordered to calculate a hash of the complete file.
 
From the link to aria2c I saw there are also "piece hashes" which is also interesting.

**Comment by nyurik** [comment]
Thanks @chridou!  Could you click the "convert to discussion" on this issue (in the right column)?

I think I might have gotten ahead of myself -- my first goal is to implement the downloader for my app -- from reading #26, it seems you don't allow HTTP downloads yet?  Or you do but it is buried deep in the core?  Do you have an example of calling API to initiate download with a given set of source HTTPs (mirrors?)

WRT piece hashes -- I think this is needed for torrent-like downloads, and won't be useful unless you want to go into that territory.

WRT `PartStream` -- if there is no current API to stream data in-order while it is being downloaded, there is really no point in using it after it is done, because i can simply use std fs functions to open and stream file content into some MD5 hashing function.

**Reply by chridou** [reply]
Hi @nyurik . The logic for HTTP is in fact not present at all. Core does not make any assumptions on the protocol being used. The only thing which needs to be implemented is the [CondowClient](https://github.com/chridou/condow/blob/main/condow_core/src/condow_client.rs#L64) trait which has just 2 functions. Nothing has to be changed in the core lib (except for checksums...). 

Examples can be found in the 2 other crates of this library:
* For [S3 with Rusoto](https://github.com/chridou/condow/blob/main/condow_rusoto/src/lib.rs#L191)
* For the [Filesystem](https://github.com/chridou/condow/blob/main/condow_fs/src/lib.rs#L52)

Up to now I only used the library for S3 downloads so I postponed HTTP because there were some open questions I did not have a quick answer for (which you are also targeting (explicit mirrors)):

* The easy case would be a user specifies a single host because he has a CDN behind that
* The user wants to specify hosts to select from explicitly. Should that be done per request or does the [CondowClient](https://github.com/chridou/condow/blob/main/condow_core/src/condow_client.rs#L64) keep a list to select from? 
* Also: Would the above be 1 or 2 implementations of the client since the first can take a complete URL while the second only paths on the defined hosts? Or even something more sophisticated?

Thanks for your interested by the way and help is of course welcome. If you have questions please feel free to ask.

**Reply by nyurik** [reply]
I don't think there should be any magical partial URL handling.  User should provide one or more complete URLs.  `condow` could verify all files are at least the same size by calling `HEAD` first, but that's optional. I think it should be a single HTTP client object of sorts that keeps track of all sources, per source retries, and possibly even cross-source balancing (i.e. one source is slower than another, so the system could do a work steal -- shortening the download from the first source, and initiating another download on another source).  This is all very similar to the logic in Rayon and other parallel task libs.

**Reply by chridou** [reply]
Regarding the streams:
There is [PartStream](https://github.com/chridou/condow/blob/main/condow_core/src/streams/part_stream.rs#L57) which returns everything ordered and [ChunkStream](https://github.com/chridou/condow/blob/main/condow_core/src/streams/chunk_stream.rs#L60) which returns the chunks as received from the network. 

Here is how the cor lib actually works:
Slice the downloads into parts and download the parts concurrently. All chunks from the parts are simply pushed onto the ChunkStream. This chunk stream is then the input for the PartStream so that parts are put into order like if the file would have been downloaded with a single requests.  It is still a stream of chunks.

I think for calculating a checksum we would need to hash the bytes in the correct order and the first "place" where we order the chunks is the PartStream.

With ChunkStream one could e.g. fill a buffer that has the size of the BLOB to be downloaded allocated. With [PartStream] a `Vec` starting empty could be filled.

**Reply by chridou** [reply]
Currently there is no work stealing. Up to now the download speed was fast enough and I was mainly targeting S3. 
I think getting this into the core lib would be a challenging and interesting tasks. We'd need some statistics and the possibility to directly access the hosts. This should be doable.

Retries are currently implemented within core since I thought this would solve this for all (upcoming) clients. But now the requirements get a bit more involved ;-)

Or let's say it has at least some "optimization" to avoid slow downloads. It creates a stream of parts to be downloaded and once a task is finished downloading a task it grabs the next from the queue. There is a small queue in front of each task which can make it prone to slow connections if the "slow host" is in many of these small queues. Could definitely be done smarter ;-)

**Reply by nyurik** [reply]
Thx, so you are thinking of making PartStream capable of reading back data from disk?   Because obviously as things arrive out of order for a huge file, things should be written to disk well ahead of when the first block may become available, and when it is, that's when it would need to stream data into the hashing function.

Regardless, seems the first order  of business is to figure out the http client -- as it is clearly the most needed one to even get to the MVP. Something like this, naming is TBD:

```rust
let cfg = HttpConfig::new("https://example.com/file.gz");
// or
let cfg = HttpConfig::new_multisource(vec!["https://example.org/file.gz", "https://another.example.org/file.gz"]);
// or
let cfg = HttpConfig::default()
            .add_source(Source::new("https://example.org/file.gz").max_streams(1)),
            .add_source(Source::new("https://another.example.org/file.gz").max_streams(2));

// There is really no need to explicitly use HttpClient::new() -- could be an internal thing
let condow = cfg.condow().unwrap();
```

One issue I see is that you have two "configuration" systems - one for the client and another for core itself. In theory, core config should not be exposed directly, but instead be a sub-object of the per-client config since client is the starting point.

The whole issue with multi-streamed download is that it has to be implemented inside the HTTP client. But this is due to the HTTP client being the starting point.  What if instead API had 3 components instead of two:  core, client, and "main" downloader.   Core defines shared interfaces, client implements it, downloader accepts one or more client instances (can even allow multiple types of clients if the interface is generic enough).

```rust
let result = condown::new(HttpClient::new("https://example.org/file.gz").max_retries(3).max_streams(2))
               .add(HttpClient::new("https://another.example.org/file.gz").max_retries(3).max_streams(1))
               .add(GcpClient::new(...))
               .save_to(PathBuf::from("./file.gz")
               .on_chunk_ordered(|chunk| { println!("got another chunk!") })  // TBD if this is the right approach
               .validate_md5("0123456789ABCDEF")  // Might as well provide a few basic ones like md5 and sha
               .run().unwrap();

**Reply by nyurik** [reply]
I don't think you need to keep any stats other than when each downloader started on its own chunk . Basic algorithm would be to:
* break file into N chunks and initiate downloads using M downloaders, where M < N
* downloaders could iterate through all initial chunks without any stealing at first
* once some downloader finishes and has nothing else to do, it could pick the slowest progressing downloader, and steal the last 40% of the remaining work (with some limits, i.e. must estimate > N seconds of download).  Performance can be computed with `downloaded_bytes / (now - start).seconds()`, so easy enough. No need to do "the speed has been slowly growing" type of thing :)

**Reply by chridou** [reply]
I see. It seems to make sense to make the internals usable as some kind of toolkit. This would require some re-engineering. So maybe we start small? 

Regarding the config for now I would simply embed the core config.
Also I see some different use case in your example. Up to now the use case was more like reusing the condow struct for the whole lifetime of the application. (I did this with the intention using this server side).

Maybe we start small with just a client that takes N URLs and chooses them round robin (maybe with some extras)? From there on we could explore the API and what we need to change and improve. Also I like the fluent API for a single download. Regarding HTTP this seems to perfectly make sense.

WDYT?

**Reply by nyurik** [reply]
@chridou i'm not sure i understood what you meant by "using it server side" -- a web server? Or are you thinking more in terms of using the condow lib in some code that runs on a server vs desktop (i.e. orchestrated via kubernetes/helm in a pod, etc)?  If later, than having a single object that coordinates a download from source(s) works the same way as for running on a desktop, wouldn't it?

WRT fluent API (i was thinking more in terms of a Rust's "builder pattern") - I think it would work just as well for non-http:  E.g. you could construct a `SomeCloudSource` object with any credentials and other options required, and once complete, pass it to the main downloader.

Another consideration is dealing with multiple downloads (i.e. get 100 files in parallel). I don't know if condow should be aware of such scenario, or if it would be the app's task to coordinate that - i.e. sharing a connection to stream files vs imitating a new one for each. Reconnecting is expensive for small files, but small files case is not condow's target (?).

**Reply by chridou** [reply]
Yes. It is running in a K8s environment. Actually there is already the idea to have some kind of resource management since there will be the requirement to manage the number of concurrent connections over multiple downloads (maybe even throttling. There are other endpoints on the same pod which must still work).
 
It would be the same as in a desktop app. The target is indeed large files.

I'd think starting with a "most simple solution" might make it easier to discuss. Then we would have something materialized to talk about and evolve on.  Maybe we should even have a call.

> WRT fluent API (i was thinking more in terms of a Rust's "builder pattern") - I think it would work just as well for non-http: E.g. you could construct a SomeCloudSource object with any credentials and other options required, and once complete, pass it to the main downloader.

This is almost how it currently works. Create an S3 client and pass it to the downloader.

**Comment by nyurik** [comment]
I did a first pass at http implementation in https://github.com/chridou/condow/pull/36.  Seems to work, but needs a bit more polishing.  Hopefully we can use it as a base to improve overall API (per previous discussion).  Thanks!

P.S. Moved restructuring ideas to #38
